File: R/spatial.R
# Spatial attack (AssessLite core spec v0.3, spec/spatial.md).
# Attacks the spatial_translation invariance (the mechanism is the same across the
# field) by leave-one-spatial-block-out: grid the coordinate space into k x k blocks
# and refit with each block removed. A material shift means some region drives the
# estimate — the mechanism is not spatially stationary.

# assign each row to a k x k grid block from quantile bins of the two coordinates
spatial_blocks <- function(x, y, k) {
  qb <- function(v) {
    br <- unique(stats::quantile(v, seq(0, 1, length.out = k + 1), na.rm = TRUE))
    if (length(br) < 2) return(rep(1L, length(v)))
    cut(v, breaks = br, include.lowest = TRUE, labels = FALSE)
  }
  bx <- qb(x); by <- qb(y)
  ifelse(is.na(bx) | is.na(by), NA_character_, paste0(bx, "_", by))
}

test_spatial_holdout <- function(audit, k = 3) {
  coords <- audit$structure$coords
  if (is.null(coords))
    stop("spatial_holdout needs declared coordinates; pass coords = c(x, y) to structural_audit()")
  d <- audit$data
  block <- spatial_blocks(d[[coords[1]]], d[[coords[2]]], k)
  rows <- list(); n_failed <- 0
  for (b in sort(unique(block[!is.na(block)]))) {
    keep <- is.na(block) | block != b
    fit <- fit_estimate(audit, d[keep, , drop = FALSE])
    if (is.null(fit)) { n_failed <- n_failed + 1; next }
    rows[[length(rows) + 1]] <- data.frame(
      label = paste0("without spatial block ", b), estimate = fit$value, se = fit$se,
      ci_low = fit$ci_low, ci_high = fit$ci_high, n = fit$n, stringsAsFactors = FALSE)
  }
  variants <- if (length(rows)) do.call(rbind, rows) else
    data.frame(label = character(), estimate = numeric(), se = numeric(),
               ci_low = numeric(), ci_high = numeric(), n = integer(), stringsAsFactors = FALSE)
  build_test_result(audit, "spatial_holdout", "spatial_translation", variants, n_failed)
}

File: python/src/assesslite/spatial.py
"""Spatial attack (AssessLite core spec v0.3, spec/spatial.md).

Attacks the spatial_translation invariance (the mechanism is the same across the
field) by leave-one-spatial-block-out: grid the coordinate space into k x k blocks
from quantile bins and refit with each block removed. Kept identical to the R engine.
"""
from __future__ import annotations

import numpy as np
import pandas as pd

from .estimator import fit_estimate
from .stability import build_test_result

_COLS = ["label", "estimate", "se", "ci_low", "ci_high", "n"]


def _bins(v: np.ndarray, k: int) -> np.ndarray:
    edges = np.unique(np.quantile(v[~np.isnan(v)], np.linspace(0, 1, k + 1)))
    if len(edges) < 2:
        return np.ones(len(v), dtype=float)
    b = np.digitize(v, edges[1:-1], right=False).astype(float)
    b[np.isnan(v)] = np.nan
    return b


def spatial_blocks(x: np.ndarray, y: np.ndarray, k: int):
    bx, by = _bins(x, k), _bins(y, k)
    out = []
    for a, b in zip(bx, by):
        out.append(None if (np.isnan(a) or np.isnan(b)) else f"{int(a)}_{int(b)}")
    return out


def test_spatial_holdout(assessment, k: int = 3) -> dict:
    coords = assessment.structure.get("coords")
    if not coords:
        raise ValueError("spatial_holdout needs declared coordinates; pass coords=(x, y) "
                         "to StructuralAudit()")
    d = assessment.data
    block = spatial_blocks(d[coords[0]].to_numpy(dtype=float),
                           d[coords[1]].to_numpy(dtype=float), k)
    block = pd.Series(block, index=d.index)
    rows, n_failed = [], 0
    for b in sorted(x for x in block.dropna().unique()):
        keep = block.isna() | (block != b)
        fit = fit_estimate(assessment.analysis, d[keep.values])
        if fit is None:
            n_failed += 1
            continue
        rows.append({"label": f"without spatial block {b}", "estimate": fit["value"],
                     "se": fit["se"], "ci_low": fit["ci_low"], "ci_high": fit["ci_high"],
                     "n": fit["n"]})
    variants = pd.DataFrame(rows, columns=_COLS)
    return build_test_result(assessment.estimate, "spatial_holdout", "spatial_translation",
                             variants, n_failed)
